### ISSUE TITLE
Feature: disable blocked thread warnings until application is started

### ISSUE THREAD

**Issue by mmindenhall** [eclipse/vert.x#1379]
**tl;dr:**  Following the "golden rule" during application startup and initialization may be unimportant.  Therefore please provide a way to disable warnings until application is fully deployed and started.

The reason I am requesting this is because my company has a large vert.x 2 ruby application (ported from rails/sinatra) that is being upgraded to vert.x 3.  We do a lot of pre-loading during application startup, and it would be difficult to make all of that asynchronous or wrapped with execute_blocking.  We're getting the multiple warnings and large stacktrace during application startup.

But during application startup, I don't mind blocking the event loop, because there's no other work the application could be doing that's being blocked.  Once verticles are started and accepting requests, I do want to see the warnings.

Therefore, I'd like to have some mechanism to disable event loop warnings/stacktraces until the application has been fully deployed.

A few ideas for how this could be done:

1.  Add more general `disableBlockedThreadWarnings` and `enableBlockedThreadWarnings` methods to `Vertx`.  Those would call new setter on `BlockedThreadChecker` to set value on a `final AtomicBoolean`.  Within the timer task `run()` method, check value before logging.  I like this solution the best. 
1.  Add a `ready()` (for example) method to `Vertx`, along with a new option to disable warnings until `ready()` is called.  Less flexible, but same effect during application startup.
1.  Extract the 4 options that affect this (`blockedThreadCheckInterval`, `maxEventLoopExecuteTime`, `maxWorkerExecuteTime`, `warningExceptionTime`) into their own class, and provide a static instance with defaults.  Provide a way to replace the instance at runtime.

**Comment by AshPlop** [comment]
I have the same issue when starting multiple microservices on the same node at the same time.
When starting them one by one, everything is fine with the startup. The microservice load within 6 sec. But when starting them all at once, they start in 26 sec, mainly due to I/O waits (loading classes takes more time, checking JCE jar signature when building the EventBus also takes a lot more time).
Currently the number of dummy microservices you can start concurrently depends on the host's performance.
The thing is that we only need extra performances at startup and we don't need that much when the microservices are running.

**Comment by erhard** [comment]
I have the same issue and as you can see in this post from 2015  https://groups.google.com/forum/#!searchin/vertx/Karger/vertx/rbvgHUDWcsE/iQspUifkAAAJ
Tim Fox said that it is very easy to fix.....
This is my second approach to use vertx (current version) with ruby and I have the same problem since 2015....it really annoying. Can I fix it myself with ruby means ?



`Thread[vert.x-eventloop-thread-3,5,main] has been blocked for 9821 ms, time limit is 2000
    io.vertx.core.VertxException: Thread blocked
    at java.io.UnixFileSystem.getBooleanAttributes0(Native Method)
    at java.io.UnixFileSystem.getBooleanAttributes(UnixFileSystem.java:242)
    at java.io.File.exists(File.java:819)
    at jnr.posix.JavaSecuredFile.exists(JavaSecuredFile.java:97)
    at org.jruby.util.RegularFileResource.exists(RegularFileResource.java:71)
    at org.jruby.runtime.load.LibrarySearcher.findFileResourceWithLoadPath(LibrarySearcher.java:169)
    at org.jruby.runtime.load.LibrarySearcher.findResourceLibrary(LibrarySearcher.java:129)
    at org.jruby.runtime.load.LibrarySearcher.findLibrary(LibrarySearcher.java:65)
    at org.jruby.runtime.load.LibrarySearcher.findBySearchState(LibrarySearcher.java:54)
    at org.jruby.runtime.load.LoadService.findLibraryBySearchState(LoadService.java:967)
    at org.jruby.runtime.load.LoadService.findFileForLoad(LoadService.java:390)
    at org.jruby.runtime.load.LoadService.smartLoadInternal(LoadService.java:489)
    at org.jruby.runtime.load.LoadService.require(LoadService.java:396)
    at org.jruby.RubyKernel.requireCommon(RubyKernel.java:963)
    at org.jruby.RubyKernel.require19(RubyKernel.java:956)
    at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(RubyKernel$INVOKER$s$1$0$require19.gen)
at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodOneOrNBlock.call(JavaMethod.java:383)
`

**Comment by vietj** [comment]
in some cases it might be a good thing that Vert.x performs some operation using a worker.

I remember I checked for Ruby and I think something was preventing that though.

**Comment by erhard** [comment]
OK. Can I filter those Messages somehow. Its very hard to find my own errors in that bulk of output.

**Comment by erhard** [comment]
I have found it yeah !!

Here my startup script for development :
Keep an eye on VERTX_OPTS !

`export VERTX_HOME="$HOME/.local/bin/vertx3"
export PATH="$PATH:$VERTX_HOME/bin"

export VERTX_OPTS='-Dvertx.options.maxEventLoopExecuteTime=100000000000'

vertx version
vertx run server_app.rb -conf dev.json`

**Comment by AshPlop** [comment]
@erhard The thing is that proceeding this way, you will never be warned when an operation takes too much time on the event loop. 
As your database grows or your input files or whathever, you might need to improve the performance of your code.
The option you used should only be activated while debugging.

**Comment by erhard** [comment]
Of course at the moment I do it while debugging. (Thanks for the hint) I would wish this feature to be able to set for the startup and for the verticles I write  itself and not as a global setting. The problem is at the moment that only at startup all that output comes up an I do not find my own error messages any more. (Now I do) Even during production that stacktrace is in my opinion a little bit oversized But this is is a fundamental thing of java to get kilometers of stack trace I think which is not changeable. A very good feature in my opinion would be if some lines of warning were logged in production in a separate log file . To see what is blocking So you can check the "blocking.log". For example (<time> vertical <name> <id> blocks <duration of blocking> The problem is here blocking is not a real error it just slows down the processing and should be avoided.

**Comment by AshPlop** [comment]
@erhard I don't know which logging implementation you are using but you can configure it to log everything comming from the BlockedThreadChecker in a separate file and limit the stacktrace to the number of lines you need to see if you code is at stake.

Here is an example using logback.

```
<property name="five_lines_stack_traces_pattern" value="%d{yyyy-MM-dd HH:mm:ss} [%-5level] %logger{36} - %msg%n%ex{5}"/>

<appender name="blocking_issues" class="ch.qos.logback.core.rolling.RollingFileAppender">
    <file>${destination}</file>
    <rollingPolicy class="ch.qos.logback.core.rolling.SizeAndTimeBasedRollingPolicy">
        <fileNamePattern>${destination}-%d{yyyy-MM-dd}.%i.txt</fileNamePattern>
        <maxFileSize>15MB</maxFileSize>
        <maxHistory>20</maxHistory>
    </rollingPolicy>
    <encoder>
        <Pattern>${five_lines_stack_traces_pattern}</Pattern>
    </encoder>
</appender>

<logger name="io.vertx.core.impl.BlockedThreadChecker" level="info" additivity="false">
    <appender-ref ref="blocking_issues"/>
</logger>

```

**Comment by erhard** [comment]
Oh great thanks a lot I am a ruby guy.....I will google howto put logback into vertx and configure it according your post. If you have some links or hints I would appreciate it.

**Comment by erhard** [comment]
@AshPlop I did what you said an have now the blocking messages in a separate rotating log file. I used logback from the startup script:

#!/usr/bin/env bash 

export VERTX_HOME="$HOME/.local/bin/vertx3"
export PATH="$PATH:$VERTX_HOME/bin"
export LOG_DESTINATION=./log

export VERTX_OPTS='-Dvertx.logger-delegate-factory-class-
name=io.vertx.core.logging.SLF4JLogDelegateFactory -Dlogback.configurationFile=file:./logback.xml'

vertx run server_app.rb -conf dev.json


Thanks for your hints

**Comment by greggwon** [comment]
Ultimately developers don't need this kind of nonsense being done for them and being in their face.  You are wasting CPU cycles by doing this continuously and it so interferes with debugging as to be useless.  Developers who don't understand performance and threading/locking/blocking issues are doomed to failure no matter what you try to do to hold their hand.  Please just disable this stupid attempt at hand holding and try something else if you really feel you are more of an expert at software that the rest of the developers own the world are.

**Comment by trajano** [comment]
I was curious what was causing it so I did

```
        final VertxOptions vertOptions = new VertxOptions();
        vertOptions.setWarningExceptionTime(1);
```

Based on the output below, I sort of like this idea because I'm getting the warning message simply starting up an HTTPS server.

```
00:11:06.746 [vertx-blocked-thread-checker] WARN  i.v.core.impl.BlockedThreadChecker - Thread Thread[vert.x-eventloop-thread-0,5,main] has been blocked for 2414 ms, time limit is 2000
io.vertx.core.VertxException: Thread blocked
	at io.netty.handler.ssl.ReferenceCountedOpenSslServerContext.newSessionContext(ReferenceCountedOpenSslServerContext.java:123)
	at io.netty.handler.ssl.OpenSslServerContext.<init>(OpenSslServerContext.java:349)
	at io.netty.handler.ssl.OpenSslServerContext.<init>(OpenSslServerContext.java:334)
	at io.netty.handler.ssl.SslContext.newServerContextInternal(SslContext.java:414)
	at io.netty.handler.ssl.SslContextBuilder.build(SslContextBuilder.java:402)
	at io.vertx.core.net.impl.SSLHelper.createContext(SSLHelper.java:295)
	at io.vertx.core.net.impl.SSLHelper.getContext(SSLHelper.java:446)
	at io.vertx.core.net.impl.SSLHelper.validate(SSLHelper.java:466)
	at io.vertx.core.http.impl.HttpServerImpl.listen(HttpServerImpl.java:255)
	at io.vertx.core.http.impl.HttpServerImpl.listen(HttpServerImpl.java:222)
	at net.trajano.ms.engine.JaxRsVerticle.start(JaxRsVerticle.java:42)
	at io.vertx.core.AbstractVerticle.start(AbstractVerticle.java:111)
	at io.vertx.core.impl.DeploymentManager.lambda$doDeploy$8(DeploymentManager.java:434)
	at io.vertx.core.impl.DeploymentManager$$Lambda$8/2141179775.handle(Unknown Source)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:337)
	at io.vertx.core.impl.ContextImpl$$Lambda$9/48914743.run(Unknown Source)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:445)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)

```

**Comment by trajano** [comment]
For the most part I would think just setting 

        final VertxOptions vertOptions = new VertxOptions();
        vertOptions.setMaxEventLoopExecuteTime(3000000000L);

Should be more than sufficient and adjust accordingly to your environment.  However, you have to make sure that it is an environment timing itself.  If it is something that does take a while I think I'd move it out of the event loop and push it into a worker when possible. (Enabling SSL in the worker isn't really something that I can be easily done from what I can tell).

**Comment by vietj** [comment]
we could for this part (SSL context creation) which can be expensive disable the the check I think.

**Comment by greggwon** [comment]
Why do you want to force the developers into discovery mode to get performant software?

If the underlying implementation is so indeterminate that you have to protect it like this, I think it's questionable in design.

Timeouts and varied latency, on a non-realtime OS, are going to happen.  Demanding something (response timing) that the user has no control over and no desire to make a requirement, is a pretty large barrier for entry...

**Comment by mmindenhall** [comment]
The issue I filed concerns the ability to change these settings at runtime, so that the _extremely_ verbose blocked thread logging can be disabled while the application is starting, and then enabled once all verticles are started.  I suggested three possible ways to accomplish this, and so far nobody has commented on pros/cons of any of those approaches (which is what I was hoping for before submitting a PR).

Would any of those approaches adequately solve the other issues that are being discussed?

**Comment by trajano** [comment]
Well for my case I put some of the larger initialization on a blocking queue and enable the route once it is fully initialized.  In my case creating a Jersey application handler can take upwards of 5 seconds so I didn't like the thread block warning message since I have it set to 1.5 second (lower than the default) with stack trace.

        vertx.executeBlocking(future -> {
            future.complete(new ApplicationHandler(resourceConfig));
        }, false,
            res -> {
                if (res.succeeded()) {
                    appHandler = (ApplicationHandler) res.result();
                    router.route(baseUri.getPath() + "*").handler(this);
                }
            });

Since I don't active the route until the requisite data is ready the code won't get a null pointer exception.

**Comment by luxiaoxun** [comment]
In Vertx.3 try this:
        VertxOptions options = new VertxOptions();
        options.setBlockedThreadCheckInterval(3*60*1000);
        Vertx.vertx(options).deployVerticle()

But actually I don't like this idea.

**Comment by syam-nd** [comment]
> The issue I filed concerns the ability to change these settings at runtime, so that the _extremely_ verbose blocked thread logging can be disabled while the application is starting, and then enabled once all verticles are started. I suggested three possible ways to accomplish this, and so far nobody has commented on pros/cons of any of those approaches (which is what I was hoping for before submitting a PR).
> 
> Would any of those approaches adequately solve the other issues that are being discussed?

I am in the same situation. This sounds like a really common use case. I am thinking of starting up the Java application normally and once my resource initialization is complete, i can initialize the vertx.

**Comment by hutchig** [comment]
Worth remembering when thinking about what BlockedThreadChecker does in the context of #1539

**Comment by st-h** [comment]
I just started to use AWS Fargate to deploy a vertx application and I suddenly see those warnings and exceptions as well.  I have never seen those warnings at startup, when launching on an ubuntu server directly or during local development. I can also start the same docker container (that is deployed on Fargate) locally and the warnings are gone. Having a way to disable the blocked thread checker on startup would be so helpful.
Even if there just would be a method to programmatically enable or disable the blocked thread checker, that would help immensely.
